metricbeat/module/system/cpu/metrics/metrics.go

```diff
@@ -0,0 +1,174 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package metrics
+
+import (
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+)
+
+// CPU manages the CPU metrics from /proc/stat
+// *BSD and and linux only use parts of these,
+// but the APIs are similar enough that this is defined here,
+// and the code that actually returns metrics to users will be OS-specific
+type CPU struct {
+	user uint64
+	nice uint64
+	sys  uint64
+	idle uint64
+	// Linux and openbsd
+	irq uint64
+	// Linux only below
+	wait    uint64
+	softIrq uint64
+	stolen  uint64
+}
```

Comment:
> EDIT: I'm still thinking of "less ugly" ways of dealing with OS-specific data, as opposed to just returning an opaque `MapStr`. I wonder if we can add OS fields to the tag structs, and then implement a `MapStr` unmarshaller that filters the values based on the OS. For example:
> 
>     type cpu struct {
>       ...
>        nice uint64 `platform:"linux,darwin",field:"nice.pct"`
>     }

A trick that you can try to leverage is that a `MapStr` can contain structs, that will be marshalled to JSON before being sent to the output. Then the members of this struct can be defined as pointers, with `omitempty`, something like this:
```
type cpu struct {
    ...
    nice *uint64 `json:"nice.pct,omitempty"`
}
```

Then each OS would fill its fields and empty ones would be omitted.

Or, a less trickier alternative could be to use pointers too, and implement the `fill*` methods as methods of this `cpu` struct, that check the filled values to know what fields to fill.